From 77bc57c6a2468d4c28f9c8e80ca20fd4e2e48dfb Mon Sep 17 00:00:00 2001
From: Sandro <devel@penguinpee.nl>
Date: Sat, 9 Nov 2024 16:59:50 +0100
Subject: [PATCH] Use FileDialog for all file/path pickers

- Use `FileDialog` throughout
- Handle whatever `GLib.Error` throws our way
- Handle it the same way regardless of dialog
---
 bottles/frontend/windows/launchoptions.py | 43 ++++++++++++++---------
 1 file changed, 27 insertions(+), 16 deletions(-)

diff --git a/bottles/frontend/windows/launchoptions.py b/bottles/frontend/windows/launchoptions.py
index adb122a3..79dbedd6 100644
--- a/bottles/frontend/windows/launchoptions.py
+++ b/bottles/frontend/windows/launchoptions.py
@@ -279,6 +279,7 @@ class LaunchOptionsDialog(Adw.Window):
                 else:
                     # something else happened...
                     logging.warning("Error selecting post-run script: %s" % error)
+                    pass
 
         dialog = Gtk.FileDialog.new()
         dialog.set_title("Select Post-run Script")
@@ -296,25 +297,35 @@ class LaunchOptionsDialog(Adw.Window):
         self.btn_post_script_reset.set_visible(False)
 
     def __choose_cwd(self, *_args):
-        def set_path(dialog, response):
-            if response != Gtk.ResponseType.ACCEPT:
-                self.action_cwd.set_subtitle(self.__default_cwd_msg)
-                return
-
-            directory_path = dialog.get_file().get_path()
-            self.program["folder"] = directory_path
-            self.action_cwd.set_subtitle(directory_path)
-            self.btn_cwd_reset.set_visible(True)
+        def set_path(dialog, result):
 
-        dialog = Gtk.FileChooserNative.new(
-            title=_("Select Working Directory"),
-            parent=self.window,
-            action=Gtk.FileChooserAction.SELECT_FOLDER,
-        )
+            try:
+                directory = dialog.select_folder_finish(result)
+
+                if directory is None:
+                    self.action_cwd.set_subtitle(
+                        self.__default_cwd_msg)
+                    return
 
+                directory_path = directory.get_path()
+                self.program["folder"] = directory_path
+                self.action_cwd.set_subtitle(directory_path)
+                self.btn_cwd_reset.set_visible(True)
+            except GLib.Error as error:
+                # also thrown when dialog has been cancelled
+                if error.code == 2:
+                    # error 2 seems to be 'dismiss' or 'cancel'
+                    if self.program["folder"] is None or self.program["folder"] == "":
+                        self.action_cwd.set_subtitle(self.__default_cwd_msg)
+                else:
+                    # something else happened...
+                    logging.warning("Error selecting folder: %s" % error)
+                    raise
+
+        dialog = Gtk.FileDialog.new()
+        dialog.set_title("Select Working Directory")
         dialog.set_modal(True)
-        dialog.connect("response", set_path)
-        dialog.show()
+        dialog.select_folder(parent=self.window, callback=set_path)
 
     def __reset_cwd(self, *_args):
         """
-- 
2.47.0

